pkg/authorization/admission/restrictnodes/restrictnodes.go

```diff
@@ -0,0 +1,222 @@
+package restrictnodes
+
+import (
+	"errors"
+	"io"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/controller/informers"
+	"k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/labels"
+
+	"k8s.io/kubernetes/pkg/admission"
+)
+
+func init() {
+	admission.RegisterPlugin("openshift.io/RestrictNodes",
+		func(kclient kclientset.Interface, config io.Reader) (admission.Interface, error) {
+			return NewRestrictNodesAdmission(kclient)
+		})
+}
+
+// restrictNodesAdmission implements admission.Interface and enforces
+// restrictions on nodes to limit their access to mutate other nodes,
+// mutate pods bound to other nodes, and to access secrets.
+type restrictNodesAdmission struct {
+	*admission.Handler
+
+	kclient kclientset.Interface
+	pods    *cache.StoreToPodLister
+}
+
+var _ admission.WantsInformerFactory = &restrictNodesAdmission{}
+
+// NewRestrictNodesAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project.
+func NewRestrictNodesAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictNodesAdmission{
+		Handler: admission.NewHandler(admission.Get, admission.Create, admission.Update, admission.Delete),
+		kclient: kclient,
+	}, nil
+}
+
+var (
+	secretsResource = kapi.Resource("secrets")
+	nodesResource   = kapi.Resource("nodes")
+	podsResource    = kapi.Resource("pods")
+)
+
+func (q *restrictNodesAdmission) Admit(a admission.Attributes) error {
+	// We only care about nodes we can identify
+	userInfo := a.GetUserInfo()
+	if userInfo == nil {
+		return nil
+	}
+	username := userInfo.GetName()
+	if !strings.HasPrefix(username, "system:node:") {
+		return nil
+	}
+	nodeName := strings.TrimPrefix(username, "system:node:")
+
+	// We only care about secrets, pods, and nodes; ignore anything else.
+	switch a.GetResource().GroupResource() {
+	case secretsResource:
+		return q.admitSecret(nodeName, a)
+	case nodesResource:
+		return q.admitNode(nodeName, a)
+	case podsResource:
+		return q.admitPod(nodeName, a)
+	default:
+		return nil
+	}
+}
+
+var (
+	everything = labels.Everything()
+)
+
+func (q *restrictNodesAdmission) admitSecret(nodeName string, a admission.Attributes) error {
+	ns := a.GetNamespace()
+	name := a.GetName()
+	pods, err := q.pods.Pods(ns).List(everything)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		if pod.Spec.NodeName != nodeName {
+			continue
+		}
+		found := false
+		visitSecretNames(pod, func(referencedName string) bool {
```

Comment:
This is what we could memo-ize if it became necessary - invert this to build a set of everything in the namespace, LRU cache it for 30 seconds (assuming all secret accesses are correlated in time), and if you get a match check one more time.  But may not even be that slow.